python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,158 @@
+"""
+Method for applying arbitrary UDFs over grouped data.
+"""
+
+from distutils.version import LooseVersion
+from itertools import chain
+from py4j import java_gateway
+
+from pyspark import SparkContext
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import Row
+
+
+def gapply(grouped_data, func, schema, *cols):
+    """WARNING: this function is planned to be migrated to apache/spark for release 2.1. This
+    API will not be maintained.
+
+    Applies the function `func` to the grouped data; in particular, by default this calls
+    `func(key1, key2, ..., keyn, values)` where the number and order of the key arguments is
+    determined by columns on which this instance's parent :class:`DataFrame` was grouped and
+    `values` is a `pandas.DataFrame` of columns selected by `cols`, in that order.
+
+    `func` is expected to return a `pandas.DataFrame` of the specified schema `schema`,
+    which should be of type :class:`StructType` (output columns are of this name and order).
+
+    If `spark.conf.get("spark.sql.retainGroupColumns")` is not `u'true'`, then `func` is
```

Comment:
since config are immutable after initialization, it means you cannot mix up with keys and without keys.
This is why I suggest the function always accept the keys, but as as single tuple object or a single object so that it is easy to ignore.